crates/krun-server/src/bin/krun-server.rs

```diff
@@ -1,32 +1,94 @@
-use std::net::TcpListener;
-use std::os::fd::AsRawFd;
-use std::panic;
-use std::process::Command;
+use std::os::unix::process::ExitStatusExt as _;
 
-use anyhow::{Context, Result};
+use anyhow::Result;
 use krun_server::cli_options::options;
-use krun_server::server::start_server;
-use nix::sys::socket::{shutdown, Shutdown};
+use krun_server::server::{start_server, State};
+use log::error;
+use tokio::net::TcpListener;
+use tokio::process::Command;
+use tokio::sync::watch;
+use tokio_stream::wrappers::WatchStream;
+use tokio_stream::StreamExt as _;
 
-fn main() -> Result<()> {
+#[tokio::main]
+async fn main() -> Result<()> {
     env_logger::init();
 
     let options = options().run();
 
-    let listener = TcpListener::bind(format!("0.0.0.0:{}", options.server_port))?;
-    let listener_fd = listener.as_raw_fd();
+    let listener = TcpListener::bind(format!("0.0.0.0:{}", options.server_port)).await?;
+    let (state_tx, state_rx) = watch::channel(State {
+        connection_idle: true,
+        child_processes: 0,
+    });
 
-    let server_thread = start_server(listener);
-
-    Command::new(&options.command)
+    let server_handle = tokio::spawn(start_server(listener, state_tx));
+    tokio::pin!(server_handle);
+    let command_status = Command::new(&options.command)
         .args(options.command_args)
-        .status()
-        .with_context(|| format!("Failed to execute command {:?}", options.command))?;
+        .status();
+    tokio::pin!(command_status);
+    let mut state_rx = WatchStream::new(state_rx);
 
-    shutdown(listener_fd, Shutdown::Both)?;
-    if let Err(err) = server_thread.join() {
-        panic::resume_unwind(err);
-    }
+    let mut server_died = false;
+    let mut command_exited = false;
 
-    Ok(())
+    loop {
+        tokio::select! {
+            res = &mut server_handle, if !server_died => {
+                // If an error is received here, accepting connections from the
+                // TCP listener failed due to non-transient errors and the
```

Comment:
TODO: This is a lie. We're logging and ignoring all connection errors at this time... We should continue on transient errors but bail out on non-transient errors.

https://book.async.rs/patterns/accept-loop